src/platform/i386/kernel.c

```diff
@@ -164,21 +166,47 @@ kmain(struct multiboot *mboot, u32_t mboot_magic, u32_t esp)
 	/* uses virtual address for VGA. should be after paging_init() */
 	vga_high_init();
 #endif
-	kern_boot_comp();
-	smp_init();
+	kern_boot_comp(0);
 	timer_init();
+	lapic_init();
 	lapic_timer_init();
+	smp_init(cores_ready);
+	bsp_ready = 1;
 	kern_boot_upcall();
+
 	/* should not get here... */
 	khalt();
 }
 
+void
+smp_kmain(void)
+{
+	volatile int cpu_id = get_cpuid();
+	struct cos_cpu_local_info *cos_info = cos_cpu_local_info();
+	printk("Initialize CPU %d\n", cpu_id);
+	tss_init(cpu_id);
+	gdt_init(cpu_id);
+	idt_init(cpu_id);
+
+	paging_init();
+	kern_boot_comp(cpu_id);
+	lapic_init();
+	lapic_timer_init();
+
+	printk("New CPU %d Booted\n", cpu_id);
+	cores_ready[cpu_id] = 1;
+	// waiting for all cored booted
+	while(bsp_ready == 0);
+	while(1);
```

Comment:
Perhaps use a comment here like /* FIXME: reason for this here */

Reply:
`while(1) ;` (space before `;`)

For the current code, I'd rather see that the `while (1) ;` be in user-space *after* the `kern_boot_upcall`.  That would be an easier space for @phanikishoreg to jump in.

Reply:
I am not sure where is the start of the code after the `kern_boot_upcall`, so I put the `while (1) ;` here. Could you guys give me the link to the code after the `kern_boot_upcall`.

Reply:
@yzcode It is here in assembly: https://github.com/gwsystems/composite/blob/ppos/src/components/lib/cos_asm_upcall.S#L46

Here in C: https://github.com/gwsystems/composite/blob/ppos/src/components/lib/cos_component.c#L141